### DISCUSSION TITLE
What is the best way to support type narrowing

### DISCUSSION THREAD

**Discussion by lucasgadams** [pydantic/pydantic#8382] [Question]
Lets say I have a model that may or may not have a field set. Then if the field is not set I create the right value for it and set it. And let's say the creation of this value is complex, involves database access or something else so that I am not going to put it into something like a default_factory. Previously I would have a class A and set the field to optional, and then I would have a class B where it is required. I take in an object A (say from a post request), and then if needed create the missing value and instantiate class B so other parts of the code know that this value now exists. For example:
```
from pydantic import BaseModel


class A(BaseModel):
    val: int | None = None


class B(A):
    val: int

```

This works perfectly fine, however after upgrading my `pyright` version, this now causes a static type error. It seems this is part of a much [larger discussion ](https://github.com/microsoft/pyright/issues/5933)on python type subclassing. The error you now get is:
```
 - error: "val" overrides symbol of same name in class "A"
    Variable is mutable so its type is invariant
      Override type "int" is not the same as base type "int | None" (reportIncompatibleVariableOverride)
1 error, 0 warnings, 0 informations 
```

So what is the proper way to do something like this? From discussions I read, it sounds like in general Generics solve this use case. But that seems like of heavy handed for this simple thing. Ideally I would want the ability to 1) narrow the type (from int | None to int) but if i 2) try to change the type that should error. Thoughts?

**Comment by ImogenBits** [comment]
In general the only way to do this is by not having either class inherit the other. To see why `B` can't be a child of `A` consider this function:
```py
def set_none(obj: A) -> None:
    obj.val = None
```
This typechecks with no issues since `A`'s val member can be set to `None`. But if `B` is a subclass of `A` then we can also pass `B` objects to this, but their val's can't be `None`!
For the other way, `A` being a subclass of `B`, look at this:
```py
def get_value(obj: A) -> int:
    return obj.val
```
This also typechecks because `A`'s value always is an `int`. But if we pass a `B` instance we might not get an `int` but `None`.

So you can't do either of this directly, but there are some workarounds that get you most of the way there! The easiest is to just make them unrelated classes:
```py
class A(BaseModel):
    val: int | None = None

class B(BaseModel):
    val: int
```
This now has two annoying things: code duplication and not having a common type to refer to either of them. The first cannot be fixed all the way, that's just a limitation of Python's type system (for example, typescript has a great solution for this problem with no code duplication). But we can mitigate it by moving all shared attributes to a common parent.
```py
class ThingyBase(BaseModel):
    something: bool
    otherthing: str

class A(BaseModel):
    val: int | None = None

class B(BaseModel):
    val: int
```
This way you only have to declare the members with different types twice. While this looks a bit like it also solves the problem of not having a common type for both, it doesn't actually do that yet. If you have a variable of type `ThingyBase` it won't let you access the val member. To solve that we just introduce a type alias, either with `type Thingy = A | B`, `Thingy = A | B`, or `Thingy: TypeAlias = A | B` depending on your python version and how much your code lets pyright infer about things. To narrow the type you can then just do the normal `isinstance` and pyright will do the rest.
